src/diag.rs

```diff
@@ -187,7 +187,7 @@ fn annotate_diagnostic(notes: &mut Vec<Notation>,
     }
 
     fn t(v: &Token) -> String {
-        String::from_utf8((**v).to_owned()).expect("utf-8 is checked before making Token")
+        unsafe { String::from_utf8_unchecked((**v).to_owned()) }
```

Comment:
Moving it to utils is also reasonable. I added this for consistency with `parser::to_str`. I've been vaguely worried about the accumulation of unnecessary string checks when converting `Vec<[u8]>` to `String` and `&[u8]` to `&str`. The rust people ensure safety by maintaining a correctness contract with the type, and we could do the same thing with `Token` / `TokenPtr` if they were newtypes with checked `new()`. We're obviously maintaining our own invariants, so this call itself is perfectly safe; the issue is with how tokens get allocated in the first place.